src/tree/tree_model.cc

```diff
@@ -17,158 +22,601 @@ DMLC_REGISTER_PARAMETER(TreeParam);
 namespace tree {
 DMLC_REGISTER_PARAMETER(TrainParam);
 }
-// internal function to dump regression tree to text
-void DumpRegTree(std::stringstream& fo,  // NOLINT(*)
-                 const RegTree& tree,
-                 const FeatureMap& fmap,
-                 int nid, int depth, int add_comma,
-                 bool with_stats, std::string format) {
-  int float_max_precision = std::numeric_limits<bst_float>::max_digits10;
-  if (format == "json") {
-    if (add_comma) {
-      fo << ",";
-    }
-    if (depth != 0) {
-      fo << std::endl;
-    }
-    for (int i = 0; i < depth + 1; ++i) {
-      fo << "  ";
-    }
-  } else {
-    for (int i = 0; i < depth; ++i) {
-      fo << '\t';
+
+/*!
+ * \brief Base class for dump model implementation, modeling closely after code generator.
+ */
+class TreeGenerator {
+ protected:
+  static int32_t constexpr kFloatMaxPrecision =
+      std::numeric_limits<bst_float>::max_digits10;
+  FeatureMap const& fmap_;
+  std::stringstream ss_;
+  bool const with_stats_;
+
+  template <typename Float>
+  static std::string ToStr(Float value) {
+    static_assert(std::is_floating_point<Float>::value,
+                  "Use std::to_string instead for non-floating point values.");
+    std::stringstream ss;
+    ss << std::setprecision(kFloatMaxPrecision) << value;
+    return ss.str();
+  }
+
+  static std::string Tabs(uint32_t n) {
+    std::string res;
+    for (uint32_t i = 0; i < n; ++i) {
+      res += '\t';
     }
+    return res;
   }
-  if (tree[nid].IsLeaf()) {
-    if (format == "json") {
-      fo << "{ \"nodeid\": " << nid
-         << ", \"leaf\": " << std::setprecision(float_max_precision) << tree[nid].LeafValue();
-      if (with_stats) {
-        fo << ", \"cover\": " << std::setprecision(float_max_precision) << tree.Stat(nid).sum_hess;
-      }
-      fo << " }";
-    } else {
-      fo << nid << ":leaf=" << std::setprecision(float_max_precision) << tree[nid].LeafValue();
-      if (with_stats) {
-        fo << ",cover=" << std::setprecision(float_max_precision) << tree.Stat(nid).sum_hess;
-      }
-      fo << '\n';
+
+  static std::string Match(std::string const& input,
+                           std::map<std::string, std::string> replacements) {
```

Comment:
Can we consider using fmtlib (https://fmt.dev/latest/index.html) for string templating? It basically lets you use Python-like formatting in C++11. I've used it extensively for dmlc/treelite and so far I've had good experience.

```cpp
std::string result = fmt::format("{foo} {bar}",
                                 "foo"_a = "hello", "bar"_a = 32)
```

Of course, every additional dependency causes more burden, so you should decide whether Python-like formatting would be sufficiently useful to worth the cost.

Reply:
If you'd like, I can quickly make an example with fmtlib.

Reply:
Thanks, I looked into treelite  and interested in GPU algorithm, but doesn't have the right now.  :)

I don't think I need to use fmt for this simple purpose, I benchmarked my code and it can generate a thousand trees within second.  

If you think the readability would be better with fmt, I can create something similar but simplified.  I am OK with dependencies, but reluctant to use git submodules for dependencies management ...

Reply:
@trivialfis I do think that using fmtlib would improve readability (due to Python-like syntax) and ease formatting for floating-point values (no call to `ToStr()` needed). But I'll defer to you as to whether that benefit is worth adding a new git submodule.